### PR TITLE
FieldLoader.GetValue() readability and maintainability improvement. 

### PR DESCRIPTION
**Pull request by GSonderling**
The original GetValue() method had over 400 lines of code, majority of which was one long `if...else` statement that also handled parsing of all values.

I created separate non-public methods for all types and replaced the `if...else` with one switch.
That could theoretically improve performance, since switches are compiled into hash tables and thus outperform very long `if...else` statements.

But measuring such improvement would be, at best, complicated and at worst a waste of time. So I have to trust compiler on this one.

The important part is, that `switch` statement is easier to read, and thus easier to debug, than old `if...else`. The same goes for methods parsing the actual values. 
It is also a lot easier to adjust and extend. Should it be necessary.

I also removed `public` access modifier from ` GetValue()` since it's only called from the same class.

### CODE REVIEW

File: OpenRA.Game/FieldLoader.cs [hunk 1]
```diff
@@ -467,202 +479,345 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
...
+				case "UncloakType":
+				case "TimeOrder":
+				case "ScrollPanelAlign":
+				case "SourceType":
+				case "BlendMode":
+				case "EnterBehaviour":
+				case "TrailType":
+				case "BeamRenderableShape":
+					return ParseEnum(fieldName, value, fieldType);
+				case "BitSet`1":
```
Comment (GraionDilach) [inline OpenRA.Game/FieldLoader.cs]: Possible typo?

**Comment by GraionDilach** [comment]
Ohwait, this explicitly lists all enums from the solution in the switch-case. That's absolutely a no-go, since Game shouldn't be aware of values or designations coming from Common or a thirdparty source.

File: OpenRA.Game/FieldLoader.cs [hunk 2]
```diff
@@ -467,202 +479,345 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
...
+				case "OwnerType":
+				case "Unit":
+				case "UncloackType":
+				case "UnitStance":
+				case "UncloakType":
+				case "TimeOrder":
+				case "ScrollPanelAlign":
+				case "SourceType":
+				case "BlendMode":
+				case "EnterBehaviour":
```
Comment (pchote) [inline OpenRA.Game/FieldLoader.cs]: Unfortunately hardcoding type names here violates the mod-engine separation.  Fieldloader *must* keep the ability to dynamically resolve enums.

File: OpenRA.Game/FieldLoader.cs (hunk 1 above)
Comment (GSonderling) [reply OpenRA.Game/FieldLoader.cs]: Looks like it. Didn't notice any related issues, but no need to leave it.

File: OpenRA.Game/FieldLoader.cs (hunk 2 above)
Comment (GSonderling) [reply OpenRA.Game/FieldLoader.cs]: If that's the case, I can put IsEnum clause back in. It shouldn't mess things up.

Comment (GSonderling) [reply OpenRA.Game/FieldLoader.cs]: Come to think of it. Shouldn't I put all the purely dynamic clauses back in? 

Now that the methods are separated, it shouldn't take long.

File: OpenRA.Game/FieldLoader.cs [hunk 3]
```diff
@@ -467,202 +479,307 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
...
+					return ParseCVectors(fieldName, value, fieldType);
+				case "BooleanExpression":
+					return ParseBoolExpression(fieldName, value, fieldType);
+				case "IntegerExpression":
+					return ParseIntegerExpression(fieldName, value, fieldType);
+				case "ImageFormat":
+					return ParseImageFormat(fieldName, value, fieldType);
+				case "Boolean":
+					return ParseBool(fieldName, value, fieldType);
+				case "HashSet`1":
```
Comment (GraionDilach) [inline OpenRA.Game/FieldLoader.cs]: Typo

File: OpenRA.Game/FieldLoader.cs [hunk 4]
```diff
@@ -467,202 +479,307 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
...
+					return ParseHashSet(fieldName, value, fieldType, field);
+				case "Dictionary`2":
```
Comment (GraionDilach) [inline OpenRA.Game/FieldLoader.cs]: Typo

File: OpenRA.Game/FieldLoader.cs [hunk 5]
```diff
@@ -467,202 +479,307 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
...
+					return ParsePairInt(fieldName, value, fieldType);
+				case "float2":
+					return ParsePairFloat(fieldName, value, fieldType);
+				case "float3":
+					return ParseTripletFloat(fieldName, value, fieldType);
+				case "Rectangle":
+					return ParseRectangle(fieldName, value, fieldType);
+				case "DateTime":
+					return ParseDateTime(fieldName, value, fieldType);
+				case "BitSet`1":
```
Comment (GraionDilach) [inline OpenRA.Game/FieldLoader.cs]: Typo.

Comment (GSonderling) [reply OpenRA.Game/FieldLoader.cs]: Are you referring to lines 747, 749 and 763? 

If so, then those are actual names of the relevant types.

Comment (GSonderling) [reply OpenRA.Game/FieldLoader.cs]: Well values of their Type.Name property.

Comment (GraionDilach) [reply OpenRA.Game/FieldLoader.cs]: You mean that the ``    `NUMBER`` suffix is part of the type?

Comment (GSonderling) [reply OpenRA.Game/FieldLoader.cs]: Yeah. It thought it looked weird, but that's actually the value of the relevant `.Name` property.

**Comment by GSonderling** [comment]
The problem is that these functions don't share all parameters. I can get them dictionary, I can even call them, but unless I know number and type of params needed I'll get exceptions. 

I could remake the whole thing with command pattern, but that is even more complicated than switch. I can adjust all functions so that they take all parameters, but that's just sloppy.

**Comment by GSonderling** [comment]
To elaborate, I can make all functions to work with variable number of arguments. Leading to worse readability. Like so...

```
static object ParseString(params object[] arguments)
		{
			if (arguments[2] != null && MemberHasTranslateAttribute[arguments[2]] && arguments[1] != null)
				return Regex.Replace(arguments[1], "@[^@]+@", m => Translate(m.Value.Substring(1, m.Value.Length - 2)), RegexOptions.Compiled);
			return arguments[1];
		}
```

Or I can have all functions take, all arguments, leading to redundancy:

**Comment by GSonderling** [comment]
I did some extra research on that, and it looks like the two options outlined above are the only viable ones.

Both `DynamicInvoke` and `ILGenerator` (like it's done in CustomHashFunctions) seem to add significant overhead, and to avoid them I would have to essentially replicate much of previous code. So I could choose between methods based on arguments and types. 

Unless someone has a better idea, please choose from options outlined in my previous comment.

**Comment by MunWolf** [comment]
I have one idea for this and I will have time in a couple of hours to have a look at the codebase and do a quick test before I share it.

Been a while since I did anything other then review in this repository.

**Comment by GSonderling** [comment]
There might be one other solution Lambdas in dictionary. Like this:

```
static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, MemberInfo field)
		{
			var value = yaml.Value;
			if (value != null) value = value.Trim();

			Dictionary<Type, Func<object>> ParseFunctions = new Dictionary<Type, Func<object>>()
			{
				{ typeof(int), () => ParseNumerics(fieldType, value, fieldName)},
				{ typeof(ushort), () => ParseNumerics(fieldType, value, fieldName)},
				{ typeof(long), () => ParseNumerics(fieldType, value, fieldName)},
				{ typeof(float), () => ParseNumerics(fieldType, value, fieldName)},
				{ typeof(decimal), () => ParseNumerics(fieldType, value, fieldName)},
				{ typeof(Color), () => ParseColor(fieldName, value, fieldType)},
				{ typeof(HSLColor), () => ParseHSLColor(fieldName, value, fieldType)},
				{ typeof(Color[]), () => ParseColors(fieldName, value, fieldType)},
				{ typeof(string), () => ParseString(fieldName, value, field)}
			};

			if (ParseFunctions.ContainsKey(fieldType))
			{
				ParseFunctions[fieldType].Invoke();
			}
```

**Comment by pchote** [comment]
The current set of if checks are simple and work.  Some of the options above are adding more complexity than this code really needs.

**Comment by GSonderling** [comment]
Ok, I will close the PR.

**Comment by pchote** [comment]
Ah, I meant to keep the parsing methods - that's a good idea - but change the switch back to the ifs.

Sorry if that didn't come across!

**Comment by GSonderling** [comment]
So should I reopen the PR, put the ifs back in and keep the methods?

**Pull request closed without merging**